numpy/testing/tests/test_utils.py

```diff
@@ -468,47 +468,152 @@ def test_report_fail_percentage(self):
-        y = x + 1
-        def failure():
-            assert_array_almost_equal_nulp(x, y, nulp=1000)
-        self.assertRaises(AssertionError, failure)
-
-    def test_complex2(self):
-        x = np.random.randn(10)
-        y = np.array(x, np.complex) + 1e-16 * np.random.randn(10)
+    def test_float64_pass(self):
+        # The number of units of least precision 
+        # In this case, use a few places above the lowest level (ie nulp=1)
+        nulp = 5
+        x = np.linspace(-20, 20, 50, dtype=np.float64)
+        x = 10**x
+        x = np.r_[-x, x]
+
+        # Addition
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp/2.
+        assert_array_almost_equal_nulp(x, y, nulp)
+
+        # Subtraction
+        epsneg = np.finfo(x.dtype).epsneg
+        y = x - x*epsneg*nulp/2.
+        assert_array_almost_equal_nulp(x, y, nulp)
+
+    def test_float64_fail(self):
+        nulp = 5
+        x = np.linspace(-20, 20, 50, dtype=np.float64)
+        x = 10**x
+        x = np.r_[-x, x]
+
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp*2.
+        self.assertRaises(AssertionError, assert_array_almost_equal_nulp, 
+                          x, y, nulp)
+
+        epsneg = np.finfo(x.dtype).epsneg
+        y = x - x*epsneg*nulp*2.
+        self.assertRaises(AssertionError, assert_array_almost_equal_nulp, 
+                          x, y, nulp)
+
+    def test_float32_pass(self):
+        nulp = 5
+        x = np.linspace(-20, 20, 50, dtype=np.float32)
+        x = 10**x
+        x = np.r_[-x, x]
+
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp/2.
+        assert_array_almost_equal_nulp(x, y, nulp)
+
+        epsneg = np.finfo(x.dtype).epsneg
+        y = x - x*epsneg*nulp/2.
+        assert_array_almost_equal_nulp(x, y, nulp)
+
+    def test_float32_fail(self):
+        nulp = 5
+        x = np.linspace(-20, 20, 50, dtype=np.float32)
+        x = 10**x
+        x = np.r_[-x, x]
+
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp*2.
+        self.assertRaises(AssertionError, assert_array_almost_equal_nulp, 
+                          x, y, nulp)
+
+        epsneg = np.finfo(x.dtype).epsneg
+        y = x - x*epsneg*nulp*2.
+        self.assertRaises(AssertionError, assert_array_almost_equal_nulp, 
+                          x, y, nulp)
+
+
+    def test_complex128_pass(self):
+        nulp = 5
+        x = np.linspace(-20, 20, 50, dtype=np.float64)
+        x = 10**x
+        x = np.r_[-x, x]
+
+        eps = np.finfo(x.dtype).eps
+        # The test condition needs to be a factor of two lower compared to
+        # pure floats because real and imaginary parts will change
+        y = x + x*eps*nulp/4.
+        # Convert to complex, use modified `y` array only once in `yi` to keep
+        # tests equivalent to non-complex ones
+        xi = x + x*1j
+        yi = y + y*1j
+        assert_array_almost_equal_nulp(xi, yi, nulp)
```

Comment:
Hah, fair enough. I did mean this as a general comment on all the complex
tests of course :-)

On Sat, Jan 10, 2015 at 5:33 PM, Charles Harris <notifications@github.com>
wrote:

> In numpy/testing/tests/test_utils.py
> <https://github.com/numpy/numpy/pull/5347#discussion-diff-22760563>:
>
> > +
> > +    def test_complex128_pass(self):
> > +        nulp = 5
> > +        x = np.linspace(-20, 20, 50, dtype=np.float64)
> > +        x = 10**x
> > +        x = np.r_[-x, x]
> > +
> > +        eps = np.finfo(x.dtype).eps
> > +        # The test condition needs to be a factor of two lower compared to
> > +        # pure floats because real and imaginary parts will change
> > +        y = x + x*eps*nulp/4.
> > +        # Convert to complex, use modified `y` array only once in `yi` to keep
> > +        # tests equivalent to non-complex ones
> > +        xi = x + x*1j
> > +        yi = y + y*1j
> > +        assert_array_almost_equal_nulp(xi, yi, nulp)
>
> Doesn't hurt, but I don't think that will make much difference because x
> is closer to x than y is ;) The best place to check that both real and
> imaginary parts are looked at is in the tests for failure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5347/files#r22760563>.
>



-- 
Nathaniel J. Smith
Postdoctoral researcher - Informatics - University of Edinburgh
http://vorpus.org